packages/js/components/src/sortable/sortable.tsx

```diff
@@ -67,43 +95,140 @@ export const Sortable = ( {
 		event: DragEvent< HTMLDivElement >,
 		index: number
 	) => {
-		if (
-			dropIndex !== null &&
-			dragIndex !== null &&
-			dropIndex !== dragIndex
-		) {
-			const nextItems = moveIndex( dragIndex, dropIndex, items );
-			setItems( nextItems as JSX.Element[] );
-			onOrderChange( nextItems );
-		}
-
-		setTimeout( () => {
-			setDragIndex( null );
-			setDropIndex( null );
-			onDragEnd( event );
-		}, THROTTLE_TIME );
+		persistItemOrder();
+		onDragEnd( event );
 	};
 
 	const handleDragOver = (
 		event: DragEvent< HTMLLIElement >,
 		index: number
 	) => {
-		const targetIndex = isBefore( event, isHorizontal ) ? index : index + 1;
+		if ( dragIndex === null ) {
+			return;
+		}
+
+		// Items before the current item cause a one off error when
+		// removed from the old array and spliced into the new array.
+		let targetIndex = dragIndex < index ? index : index + 1;
+		if ( isBefore( event, isHorizontal ) ) {
+			targetIndex--;
+		}
+
 		setDropIndex( targetIndex );
 		onDragOver( event );
 	};
 
 	const throttledHandleDragOver = useCallback(
 		throttle( handleDragOver, THROTTLE_TIME ),
-		[]
+		[ dragIndex ]
 	);
 
+	const handleKeyDown = (
+		event: KeyboardEvent< HTMLLIElement >,
+		index: number
+	) => {
+		const { key } = event;
+		const isSelecting = dragIndex === null || dropIndex === null;
+		const selectedLabel = getItemName( ref.current, selectedIndex );
+
+		if ( key === ' ' ) {
+			if ( isSelecting ) {
+				speak(
+					sprintf(
+						/** Translators: Selected item label */
+						__(
+							'%s selected, use up and down arrow keys to reorder',
+							'woocommerce'
+						),
+						selectedLabel
+					),
+					'assertive'
+				);
+				setDragIndex( selectedIndex );
+				setDropIndex( selectedIndex );
+				return;
+			}
+
+			setSelectedIndex( dropIndex );
+			speak(
+				sprintf(
+					/* translators: %1$s: Selected item label, %2$d: Current position in list, %3$d: List total length */
+					__(
+						'%1$s dropped, position in list: %2$d of %3$d',
+						'woocommerce'
+					),
+					selectedLabel,
+					dropIndex + 1,
+					items.length
+				),
+				'assertive'
+			);
+			persistItemOrder();
+		}
+
+		if ( key === 'ArrowUp' ) {
```

Comment:
(Optional) You could add `return` to these handlers for a very slight optimization, since there's no more logic. Or use a `switch` statement.